pkg/glusterblock/controllerserver.go

```diff
@@ -0,0 +1,316 @@
+	*GfDriver
+}
+
+// CsiDrvParam stores csi driver specific request parameters.
+// This struct will be used to gather specific fields of CSI driver:
+// For eg. csiDrvName, csiDrvVersion..etc
+// and also gather parameters passed from SC which not part of gluster volcreate api.
+// GlusterCluster - The resturl of gluster cluster
+// GlusterUser - The gluster username who got access to the APIs.
+// GlusterUserToken - The password/token of glusterUser to connect to glusterCluster
+// GlusterVersion - Says the version of the glustercluster running in glusterCluster endpoint.
+// All of these fields are optional and can be used if needed.
+type CsiDrvParam struct {
+	GlusterCluster   string
+	GlusterUser      string
+	GlusterUserToken string
+	GlusterVersion   string
+	CsiDrvName       string
+	CsiDrvVersion    string
+}
+
+// ProvisionerConfig is the combined configuration of gluster cli vol create request and CSI driver specific input
+type ProvisionerConfig struct {
+	gdVolReq *api.VolCreateReq
+	//csiConf  *CsiDrvParam
+}
+
+// ParseCreateVolRequest parse incoming volume create request and fill ProvisionerConfig.
+func (cs *ControllerServer) ParseCreateVolRequest(req *csi.CreateVolumeRequest) (*ProvisionerConfig, error) {
+
+	var reqConf ProvisionerConfig
+	var gdReq api.VolCreateReq
+
+	reqConf.gdVolReq = &gdReq
+
+	// Get Volume name
+	if req != nil {
+		reqConf.gdVolReq.Name = req.Name
+	}
+	return &reqConf, nil
+}
+
+// CreateVolume creates and starts the volume
+func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) { //nolint: gocyclo
+	glog.V(2).Infof("request received %+v", req)
+
+	if err := cs.validateCreateVolumeReq(req); err != nil {
+		return nil, err
+	}
+
+	glog.V(1).Infof("creating volume with name %s", req.Name)
+
+	volSizeBytes := cs.getVolumeSize(req)
+
+	// parse the request.
+	parseResp, err := cs.ParseCreateVolRequest(req)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "failed to parse request")
+	}
+
+	volumeName := parseResp.gdVolReq.Name
+
+	blockVolInfo := bapi.BlockVolumeInfo{
+		Name: volumeName,
+		Size: uint64(volSizeBytes),
+	}
+	volumeReq := bapi.BlockVolumeCreateRequest{
+		BlockVolumeInfo: &blockVolInfo,
+	}
+
+	glog.V(2).Infof("block volume create request: %+v", volumeReq)
+	blockVolCreateResp, err := cs.client.BlockVolumeCreate("gluster-loopback", volumeReq)
+	if err != nil {
+		glog.Errorf("failed to create block volume: %s err: %v", volumeName, err)
+		errResp := cs.client.LastErrorResponse()
+		// errResp will be nil in case of `No route to host` error
+		if errResp != nil && errResp.StatusCode == http.StatusConflict {
+			return nil, status.Errorf(codes.Aborted, "block volume create already in process: %v", err)
+		}
+
+		return nil, status.Errorf(codes.Internal, "failed to create block volume: %s err: %v", volumeName, err)
+	}
+
+	glusterServer, bkpServers, err := utils.GetClusterNodes(cs.client)
```

Comment:
do we need bkpservers also?

Reply:
It's get showed in VolumeAttributes. It's get show cluster nodes I believe.